keps/sig-node/20200619-provisioning-resources-with-numa-topology.md

```diff
@@ -0,0 +1,240 @@
+  - "@derekwaynecarr"
+creation-date: 2020-06-19
+last-updated: 2020-06-19
+status: implementable
+see-also:
+  - "/keps/sig-scheduling/20200612-deducted-topology-manager.md"
+---
+
+# Exposure Node Resources With NUMA Topology Information
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+- [Design Details](#design-details)
+  - [Design based on CRI](#cri)
+  - [Design based on podresources interface of the kubelet](#podresources)
+  - [API](#api)
+  - [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+- [Alternatives](#alternatives)
+  - [Annotation approach](#annotation-approach)
+  - [NUMA specification in ResourceName](#numa-in-resourcename)
+<!-- /toc -->
+
+## Summary
+
+Cluster which contains several nodes with NUMA topology and
+enabled TopologyManager feature gate is not rare now. In such cluster
+could be a situation when TopologyManager's admit handler on the kubelet
+side refuses pod launching since pod's resources request doesn't sutisfy
+selected TopologyManager policy, in this case pod should be rescheduled.
+And so it can get stuck in the rescheduling cycle.
+To avoid such kind of problem scheduler should choose a node considering topology
+of the node and TopologyManager policy on the worker node.
+
+## Motivation
+
+For the sake of scheduling which is topology aware, resources with topology
+information should be provisioned.
+This KEP describes how it would be implemented.
+
+### Goals
+
+Provisioning resources with topology information.
+
+### Non-Goals
+
+ - modification of any public API
+ - improving and as a result modification of the TopologyManager and its policies
+
+## Proposal
+
+Introduce standalone daemon which will be run on each node in the cluster as daemonset. It should
+collect resources for running pods and provide it with CRD.
+
+## Design Details
+
+The design consists of part which describes how datum collected
+and how it was provided.
+
+Resources used by the pod could be obtained by podresources interface
+of the kubelet and by CRI from container runtime.
+
+To calculate available resources need to know all resources
+which could be used by kubernetes. It could be calculated by
+substracting resources of kube cgroup and system cgroup from
+system allocatable resources.
+
+### Design based on CRI
+
+The containerStatusResponse returned as a response to the ContainerStatus rpc contains `Info` field which is used by the container runtime for capturing ContainerInfo.
+```go
+message ContainerStatusResponse {
+      ContainerStatus status = 1;
+      map<string, string> info = 2;
+}
+```
+
+Containerd has been used as the container runtime in the initial investigation. The internal container object info
+[here](https://github.com/containerd/cri/blob/master/pkg/server/container_status.go#L130)
+
+The Daemon set is responsible for the following:
```

Comment:
this only works if the container is running, right?

if the container is in crashloop backoff, or is unable to start right away (lets say the volume has not yet mounted), by introspecting the CRI state, you get an incomplete view.  have you thought about whether the kubelet itself should expose an endpoint?

Reply:
We've considered two ways of obtaining topology information CRI and podresources. Bot interface should be extended to do it.

Reply:
Your assessment here is correct, we check for resources allocated to running pods. We haven't explored kubelet exposing an endpoint yet.